frontend/js/src/common/UserSearch.tsx

```diff
@@ -14,97 +11,131 @@ export type UserSearchProps = {
 
-export default class UserSearch extends React.Component<
-  UserSearchProps,
-  UserSearchState
-> {
-  static contextType = GlobalAppContext;
-  declare context: React.ContextType<typeof GlobalAppContext>;
+  // Props
+  const { onSelectUser, placeholder, clearOnSelect, initialValue } = props;
 
-  // eslint-disable-next-line react/sort-comp
-  constructor(props: UserSearchProps) {
-    super(props);
-    this.state = {
-      newUser: props.initialValue ?? "",
-      userSearchResults: [],
-    };
-  }
+  // States
+  const [newUser, setNewUser] = React.useState(initialValue ?? "");
+  const [userSearchResults, setUserSearchResults] = React.useState<
+    Array<SearchUser>
+  >([]);
+  const [selectedIndex, setSelectedIndex] = React.useState(-1);
 
-  throttledSearchUsers = _throttle(async () => {
-    await this.searchUsers();
-  }, 300);
+  // Refs
+  const dropdownRef = React.useRef<HTMLSelectElement>(null);
 
-  searchUsers = async () => {
-    const { APIService } = this.context;
-    const { newUser } = this.state;
+  const searchUsers = async () => {
     try {
       const response = await APIService.searchUsers(newUser);
-      this.setState({
-        userSearchResults: response.users,
-      });
+      setUserSearchResults(response.users);
     } catch (error) {
-      this.handleError(error);
+      toast.error(<ToastMsg title="Error" message={error.message} />, {
+        toastId: "error",
+      });
     }
   };
 
-  handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
-    this.setState(
-      {
-        newUser: event.target.value,
-      },
-      this.throttledSearchUsers
-    );
-  };
+  const throttledSearchUsers = _throttle(async () => {
+    await searchUsers();
+  }, 300);
 
-  handleResultClick = (user: string) => {
-    const { onSelectUser, clearOnSelect } = this.props;
+  const handleResultClick = (user: string) => {
     onSelectUser(user);
-    this.setState({
-      newUser: clearOnSelect ? "" : user,
-      userSearchResults: [],
-    });
+    setNewUser(clearOnSelect ? "" : user);
+    setUserSearchResults([]);
   };
 
-  handleError = (error: any) => {
-    toast.error(<ToastMsg title="Error" message={error.message} />, {
-      toastId: "error",
-    });
+  const reset = () => {
+    setUserSearchResults([]);
+    setSelectedIndex(-1);
   };
 
-  render() {
-    const { newUser, userSearchResults } = this.state;
-    const { placeholder } = this.props;
-    return (
-      <>
-        <input
-          id="user-name-search"
-          type="text"
-          className="form-control"
-          name="newUser"
-          onChange={this.handleInputChange}
-          placeholder={placeholder}
-          value={newUser}
-          aria-haspopup={Boolean(userSearchResults?.length)}
-        />
-        <div className="search-dropdown">
-          {userSearchResults?.map((user) => {
+  const handleKeyDown = (event: React.KeyboardEvent) => {
```

Comment:
Thanks for adding the accessibility improvements here too !

If we run into it again it might make sense to create a custom component for selects that implement these mechanisms, to be reused.

Or maybe something along the lines of this hook? https://github.com/dealroom/react-arrow-key-navigation-hook

Reply:
This article might be interesting: https://css-tricks.com/striking-a-balance-between-native-and-custom-select-elements/

I ended up there while trying to figure out something about selects while working on #2894